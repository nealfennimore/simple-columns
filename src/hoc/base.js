import { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class Base extends PureComponent {
    static propTypes = {
        className: PropTypes.string, // eslint-disable-line react/require-default-props
    }

    get entries() {
        return Object.entries( this.props );
    }

    get derivedClassNames() {
        return this.entries.reduce( ( acc, [key, value] )=> {
            if( value && this.keys.includes( key ) ) {
                const className = this.mapping[key];
                acc.push( className );
            }
            return acc;
        }, [] );
    }

    get className() {
        return [ 'flex', ...this.derivedClassNames, this.props.className]
            .reduce( ( acc, className )=> {
                if( className && typeof className === 'string' ) {
                    acc.push( className.trim() );
                }
                return acc;
            }, [] )
            .join( ' ' );
    }

    get passthroughProps() {
        return this.entries.reduce( ( acc, [key, value] ) => {
            if( ! this.keys.includes( key ) ) {
                acc[key] = value;
            }
            return acc;
        }, {} );
    }
}
