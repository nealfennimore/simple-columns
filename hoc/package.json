{
    "name": "simple-columns/hoc",
    "private": true,
    "main": "../lib/hoc/index.js",
    "module": "../es/hoc/index.js",
    "jsnext:main": "../es/hoc/index.js"
}