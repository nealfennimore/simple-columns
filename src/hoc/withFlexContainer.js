import React from 'react';
import PropTypes from 'prop-types';
import Base from './base';
import './withFlexContainer.pcss';

export const flexContainerMapping = {
    // -------------
    // flex-direction
    // -------------
    directionRow: 'flex-direction--row',
    directionRowReverse: 'flex-direction--row-reverse',
    directionColumn: 'flex-direction--column',
    directionColumnReverse: 'flex-direction--column-reverse',
    // -------------
    // flex-wrap
    // -------------
    noWrap: 'flex-wrap--nowrap',
    wrap: 'flex-wrap--wrap',
    wrapReverse: 'flex-wrap--wrap-reverse',
    // -------------
    // justify-content
    // -------------
    justifyStart: 'justify-content--start',
    justifyEnd: 'justify-content--end',
    justifyCenter: 'justify-content--center',
    justifyBetween: 'justify-content--space-between',
    justifyAround: 'justify-content--space-around',
    justifyEvenly: 'justify-content--space-evenly',
    // -------------
    // align-items
    // -------------
    itemsStart: 'align-items--start',
    itemsEnd: 'align-items--end',
    itemsCenter: 'align-items--center',
    itemsBaseline: 'align-items--baseline',
    itemsStretch: 'align-items--stretch',
    // -------------
    // align-content
    // -------------
    contentStart: 'align-content--space-start',
    contentEnd: 'align-content--space-end',
    contentCenter: 'align-content--space-center',
    contentBetween: 'align-content--space-between',
    contentAround: 'align-content--space-around',
    contentEvenly: 'align-content--space-event',
    contentStetch: 'align-content--space-stretch',
};

export const flexContainerKeys = Object.keys( flexContainerMapping );

const flexContainerPropTypes = flexContainerKeys.reduce( ( acc, name )=>{
    acc[name] = PropTypes.bool;
    return acc;
}, {} );

const flexContainerDefaultProps = flexContainerKeys.reduce( ( acc, name )=>{
    acc[name] = false;
    return acc;
}, {} );

const withFlexContainerHOC = ( WrappedComponent ) =>{

    class FlexContainer extends Base {
        static propTypes = {
            className: PropTypes.string,
            ...flexContainerPropTypes,
        }

        static defaultProps = {
            className: '',
            ...flexContainerDefaultProps,
        }

        keys = flexContainerKeys;
        mapping = flexContainerMapping;

        render() {
            return (
                <WrappedComponent {...this.passthroughProps} className={this.className} />
            );
        }
    }

    return FlexContainer;
};

export default withFlexContainerHOC;
