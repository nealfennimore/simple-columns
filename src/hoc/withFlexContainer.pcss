.flex {
    display: flex;
}

.flex-direction--row {
    flex-direction: row;
}
.flex-direction--row-reverse {
    flex-direction: row-reverse;
}
.flex-direction--column {
    flex-direction: column;
}
.flex-direction--column-reverse {
    flex-direction: column-reverse;
}

.flex-wrap--nowrap {
    flex-wrap: nowrap;
}
.flex-wrap--wrap {
    flex-wrap: wrap;
}
..flex-wrap--wrap-reverse {
    flex-wrap: wrap-reverse;
}

.justify-content--start {
    justify-content: flex-start;
}
.justify-content--end {
    justify-content: flex-end;
}
.justify-content--center {
    justify-content: center;
}
.justify-content--space-between {
    justify-content: space-between;
}
.justify-content--space-around {
    justify-content: space-around;
}
.justify-content--space-evenly {
    justify-content: space-evenly;
}

.align-items--start {
    align-items: flex-start;
}
.align-items--end {
    align-items: flex-end;
}
.align-items--center {
    align-items: center;
}
.align-items--baseline {
    align-items: baseline;
}
.align-items--stretch {
    align-items: stretch;
}

.align-content--start {
    align-content: flex-start;
}
.align-content--end {
    align-content: flex-end;
}
.align-content--center {
    align-content: center;
}
.align-content--space-between {
    align-content: space-between;
}
.align-content--space-around {
    align-content: space-around;
}
.align-content--space-evenly {
    align-content: space-evenly;
}
.align-content--stretch {
    align-content: stretch;
}
