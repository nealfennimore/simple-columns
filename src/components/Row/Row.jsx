import Base from '../Base';
import withFlexContainer from '../../hoc/withFlexContainer';

export default withFlexContainer( Base );
