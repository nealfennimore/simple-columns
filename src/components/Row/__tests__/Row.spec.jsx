import React from 'react';
import renderer from 'react-test-renderer';
import { flexContainerKeys } from '../../../hoc/withFlexContainer';
import Row from '../Row';

describe( 'Row', ()=>{
    test( 'should match snapshot', ()=>{
        const component = renderer.create(
            <Row />
        );
        expect( component.toJSON() ).toMatchSnapshot();
    } );
    test( 'should match snapshot with children', ()=>{
        const component = renderer.create(
            <Row>children</Row>
        );
        expect( component.toJSON() ).toMatchSnapshot();
    } );
    test( 'should match snapshot with flex parent class names', ()=>{
        const props = flexContainerKeys.reduce( ( acc, prop )=>{
            acc[prop] = true;
            return acc;
        }, {} );
        const component = renderer.create(
            <Row {...props} />
        );
        expect( component.toJSON() ).toMatchSnapshot();
    } );
    test( 'should match snapshot with passthrough attributes', ()=>{
        const component = renderer.create(
            <Row className="custom" id="id" />
        );
        expect( component.toJSON() ).toMatchSnapshot();
    } );
} );
