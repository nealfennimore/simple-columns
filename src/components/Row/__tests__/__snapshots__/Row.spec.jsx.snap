// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row should match snapshot 1`] = `
<div
  className="flex"
/>
`;

exports[`Row should match snapshot with children 1`] = `
<div
  className="flex"
>
  children
</div>
`;

exports[`Row should match snapshot with flex parent class names 1`] = `
<div
  className="flex flex-direction--row flex-direction--row-reverse flex-direction--column flex-direction--column-reverse flex-wrap--nowrap flex-wrap--wrap flex-wrap--wrap-reverse justify-content--start justify-content--end justify-content--center justify-content--space-between justify-content--space-around justify-content--space-evenly align-items--start align-items--end align-items--center align-items--baseline align-items--stretch align-content--space-start align-content--space-end align-content--space-center align-content--space-between align-content--space-around align-content--space-event align-content--space-stretch"
/>
`;

exports[`Row should match snapshot with passthrough attributes 1`] = `
<div
  className="flex custom"
  id="id"
/>
`;
